torch/_ops.py

```diff
@@ -25,6 +25,109 @@ def dl_open_guard():
     if _SET_GLOBAL_FLAGS:
         sys.setdlopenflags(old_flags)
 
+# Each OpOverload object contains pointer to a a specific operator overload, a pointer to the parent `OpOverloadPacket` object.
+# You can obtain an OpOverload object through attribute query on OpOverloadPacket.
+class OpOverload:
+    def __init__(self, overloadpacket, op, schema):
+        self._op = op
+        self._schema = schema
+        self._overloadpacket = overloadpacket
+
+    # it's a no-op since OpOverload object is immutable and must be for a given op overload.
+    def __deepcopy__(self, memo=None):
+        return self
+
+    def __str__(self):
+        return "OpOverload(op='{}.{}', overload='{}')".format(*self._schema.name.split("::"), self.overload_name)
+
+    def __call__(self, *args, **kwargs):
+        return self._op(*args, **kwargs or {})
+
+    def __getattr__(self, key):
+        return getattr(self._op, key)
+
+    # `my_namespace::my_op`
+    @property
+    def name(self):
+        return "{}.{}".format(*self._schema.name.split("::"))
+
+    @property
+    def overload_name(self):
+        return self._schema.overload_name
+
+    @property
+    def overload_packet(self):
+        return self._overloadpacket
+
+    @property
+    def op(self):
+        return self._op
+
+    # TODO: add more methods to expose information about input and output arguments
+
+# OpOverloadPacket class contains pointer to a base unresolved operator that doesn't correspond to a specific operator
+# You can obtain an OpOverload object through attribute query.
+class OpOverloadPacket:
+    def __init__(self, qualified_op_name, op_name, op):
+        # These attributes are accessible on the object through the properties
+        # defined below but are immutable
+        self._qualified_op_name = qualified_op_name
+        self._op_name = op_name
+        self._op = op
+
+    # it's a no-op since OpOverloadPacket object is immutable and must be for a given op.
+    def __deepcopy__(self, memo=None):
+        return self
+
+    def __str__(self):
+        return "OpOverloadPacket(op='{}.{}')".format(*self._qualified_op_name.split("::"))
+
+    @property
+    def qualified_op_name(self):
+        return "{}.{}".format(*self._qualified_op_name.split("::"))
+
+    @property
+    def op_name(self):
+        return self._op_name
+
+    @property
+    def op(self):
+        return self._op
+
+    def __getattr__(self, key):
+        # It is not a valid op_name when __file__ is passed in
+        if key == '__file__':
+            return 'torch.ops'
+
+        try:
+            use_key = '' if key == 'default' else key
+            # TODO: disallow access to overloads registered by JIT
+            op_ = torch._C._get_operation_overload(self._qualified_op_name, use_key)
+            schema = torch._C._get_schema(self._qualified_op_name, use_key)
+            overload = OpOverload(self, op_, schema)
+            # cache the overload object
+            setattr(self, key, overload)
```

Comment:
This makes me nervous. OpOverload contains a pointer to the OpOverloadPacket (self) that it was constructed from. This looks like a reference cycle to me. Luckily it doesn't matter too much, because ops don't really get GC'ed, but it gives me bad vibes.

Reply:
is the issue here that we are giving up on instant deletion because of the ref cycle? would you suggest not having a pointer from the opoverload to opoverloadpacket?